File: KPageView/KContentView.m
//
//  KContentView.m
//  KPageView
//
//  Created by Kratos on 2017/6/28.
//  Copyright © 2017年 Kratos. All rights reserved.
//

#import "KContentView.h"
#import "KPageStyle.h"
#import "UIView+KExtension.h"
@interface KContentView ()<UICollectionViewDelegate,UICollectionViewDataSource>


@property (nonatomic ,strong) KPageStyle *style;
@property (nonatomic ,strong) NSArray *childVcs;
@property (nonatomic ,strong) NSArray *title;
@property (nonatomic ,strong) UIViewController *parentVc;
@property (nonatomic ,assign) CGFloat startOffsetX;
//禁止代理
@property (nonatomic ,assign) BOOL isForbidDelegate;
@property (nonatomic ,strong) UICollectionView *collectionView;

@end

@implementation KContentView
static  NSString *kContentCellID = @"kContentCellID";

- (instancetype)initWithFrame:(CGRect)frame ChildVcs:(NSArray *)vcs parentVc:(UIViewController *)parentVc
{
    if (self = [super initWithFrame:frame]) {
        self.childVcs = vcs;
        self.parentVc = parentVc;
        [self setupUI];
    }
    return self;
}


- (void)setupUI
{
    [self addSubview:self.collectionView];
    for (UIViewController *childVc  in self.childVcs) {
        [self.parentVc addChildViewController:childVc];
    }
}

- (UICollectionView *)collectionView
{
    if (_collectionView == nil) {
       UICollectionViewFlowLayout *layout = [[UICollectionViewFlowLayout alloc]init];
        layout.itemSize = self.bounds.size;
        layout.minimumLineSpacing = 0;
        layout.minimumInteritemSpacing = 0;
        layout.scrollDirection = UICollectionViewScrollDirectionHorizontal;
        
        _collectionView = [[UICollectionView alloc]initWithFrame:self.bounds collectionViewLayout:layout];
        _collectionView.showsHorizontalScrollIndicator = NO;
        _collectionView.pagingEnabled = YES;
        _collectionView.scrollsToTop = NO;
        _collectionView.dataSource = self;
        _collectionView.delegate = self;
        _collectionView.bounces = NO;
        [_collectionView registerClass:[UICollectionViewCell class] forCellWithReuseIdentifier:kContentCellID];
        
    }
    return _collectionView;
}

#pragma mark UICollectionViewDataSource、UICollectionViewDelegate

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
{
    return self.childVcs.count;
}


- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    UICollectionViewCell *cell =[collectionView dequeueReusableCellWithReuseIdentifier:kContentCellID forIndexPath:indexPath];
    for (UIView *view in cell.contentView.subviews) {
        [view removeFromSuperview];
    }
    UIViewController *childVc = self.childVcs[indexPath.item];
    childVc.view.frame = cell.contentView.bounds;
    [cell.contentView addSubview:childVc.view];
    return cell;
}

- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView
{
    [self collectionViewDidEndScroll];
}

- (void)scrollViewDidEndDragging:(UIScrollView *)scrollView willDecelerate:(BOOL)decelerate
{
    if (!decelerate)
    {
        [self collectionViewDidEndScroll];
    }
}
- (void)collectionViewDidEndScroll
{
    NSInteger inIndex = self.collectionView.contentOffset.x / self.collectionView.width;
    if (self.delegate && [self.delegate respondsToSelector:@selector(contentView:inIndex:)])
    {
        [self.delegate contentView:self inIndex:inIndex];
    }

}
- (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
{
    self.isForbidDelegate = NO;
    self.startOffsetX = scrollView.contentOffset.x;
}

- (void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    if (scrollView.contentOffset.x == self.startOffsetX || _isForbidDelegate) {
        return;
    }
    CGFloat progress = 0;
    NSInteger targetIndex = 0;
    NSInteger sourceIndex = (self.startOffsetX / self.collectionView.width);
    //判断用户是左滑动还是右滑动
    if (scrollView.contentOffset.x > self.startOffsetX) { //左滑动
        targetIndex = sourceIndex + 1;
        progress =  (self.collectionView.contentOffset.x - self.startOffsetX) / self.collectionView.width;
    }else
    {
        targetIndex = sourceIndex - 1;
        progress = (self.startOffsetX - self.collectionView.contentOffset.x) / self.collectionView.width;

    }
    [self.delegate contentView:self sourceIndex:sourceIndex targetIndex:targetIndex progress:progress];

    
}
#pragma mark KTitleViewDelegate
- (void)titleView:(KTitleView *)titleView currentIndex:(NSInteger)index
{
    self.isForbidDelegate = YES;
    NSIndexPath *indexPath = [NSIndexPath indexPathForItem:index inSection:0];
    [self.collectionView scrollToItemAtIndexPath:indexPath atScrollPosition:UICollectionViewScrollPositionLeft animated:NO];
}
@end

File: KPageViewSwift/KPageView/HYPageView/KContentView.swift
//
//  KContentView.swift
//  KPageView
//
//  Created by Kratos on 17/4/8.
//  Copyright © 2017年 Kratos. All rights reserved.
//

import UIKit

protocol KContentViewDelegate : class {
    func contentView(_ contentView : KContentView, inIndex : Int)
    func contentView(_ contentView : KContentView, sourceIndex : Int, targetIndex : Int, progress : CGFloat)
}

private let kContentCellID = "kContentCellID"

class KContentView: UIView {
    
    // MARK: 定义属性
    weak var delegate : KContentViewDelegate?
    fileprivate var childVcs : [UIViewController]
    fileprivate var parentVc : UIViewController
    fileprivate var startOffsetX : CGFloat = 0
    fileprivate var isForbidDelegate : Bool = false
    
    
    fileprivate lazy var collectionView : UICollectionView = {
        let layout = UICollectionViewFlowLayout()
        layout.itemSize = self.bounds.size
        layout.minimumLineSpacing = 0
        layout.minimumInteritemSpacing = 0
        layout.scrollDirection = .horizontal;
        let collectionView = UICollectionView(frame: self.bounds, collectionViewLayout: layout)
        collectionView.showsHorizontalScrollIndicator = false
        collectionView.isPagingEnabled = true
        collectionView.scrollsToTop = false
        collectionView.dataSource = self
        collectionView.delegate = self
        collectionView.bounces = false
        collectionView.register(UICollectionViewCell.self, forCellWithReuseIdentifier: kContentCellID)
        return collectionView
    }()
    
    // MARK: 构造函数
    init(frame: CGRect, childVcs : [UIViewController], parentVc : UIViewController) {
        self.childVcs = childVcs
        self.parentVc = parentVc
        super.init(frame: frame)
        setupUI()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented");
        
    }

}


extension KContentView {
    fileprivate func setupUI() {
     
        // 1.添加collectionView
        addSubview(collectionView)
        // 2.将所有的子控制器添加到父控制器中
        for childVc in childVcs {
            parentVc.addChildViewController(childVc)
        }
    }
}


extension KContentView : UICollectionViewDataSource {
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return childVcs.count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        // 1.获取cell
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: kContentCellID, for: indexPath)
        
        // 2.给cell设置内容
        for subview in cell.contentView.subviews {
            subview.removeFromSuperview()
        }
        let childVc = childVcs[indexPath.item]
        childVc.view.frame = cell.contentView.bounds
        cell.contentView.addSubview(childVc.view)
        
        return cell
    }
}


extension KContentView : UICollectionViewDelegate {
    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        collectionViewDidEndScroll()
    }
    
    func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
        if !decelerate {
            collectionViewDidEndScroll()
        }
    }
    
    func collectionViewDidEndScroll() {
        // 1.获取位置
        let inIndex = Int(collectionView.contentOffset.x / collectionView.bounds.width)
        
        // 2.通知代理
        delegate?.contentView(self, inIndex: inIndex)
    }
    
    func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
        isForbidDelegate = false
        startOffsetX = scrollView.contentOffset.x
    }
    
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        
        // 1.判断是否需要执行后续的代码
        if scrollView.contentOffset.x == startOffsetX || isForbidDelegate {
            return
        }
        
        // 2.定义需要的参数
        var progress : CGFloat = 0
        var targetIndex = 0;
        var sourceIndex = 0;
        
        progress = scrollView.contentOffset.x.truncatingRemainder(dividingBy: scrollView.bounds.width) / scrollView.bounds.width;

        if progress == 0
        {
            return;
        }
        
        let index = Int(scrollView.contentOffset.x / scrollView.bounds.width);
        
        // 3.判断用户是左滑动还是右滑动
        if collectionView.contentOffset.x > startOffsetX { // 左滑动
            sourceIndex = index
            targetIndex = index + 1
            if targetIndex > childVcs.count - 1
            {
                    return;
            }
        } else { // 右滑动
            sourceIndex = index + 1;
            targetIndex = index;
            progress = 1 - progress;
            if  targetIndex < 0 {
                return;
            }
        }

        
        // 4.通知代理
        delegate?.contentView(self, sourceIndex: sourceIndex, targetIndex: targetIndex, progress: progress)
    }
}


extension KContentView : KTitleViewDelegate {
    func titleView(_ titleView: KTitleView, currentIndex: Int) {
        // 0.设置isForbidDelegate属性为true
        isForbidDelegate = true
        
        // 1.根据currentIndex获取indexPath
        let indexPath = IndexPath(item: currentIndex, section: 0)
        
        // 2.滚动到正确的位置
        collectionView.scrollToItem(at: indexPath, at: .left, animated: false)
    }
}










